### PR TITLE
Better formatting for exception messages

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Related to https://github.com/ClickHouse/ClickHouse/pull/45342

Use `fmt::format` (and variadic ctor of `DB::Exception`) instead of string concatenations to format exception messages.

The changes were done with a bunch of trash-scripts (too trash to add to the repository) and further manual editing
https://pastila.nl/?001a6151/05fa0d1fba0f199d9b560674a6d6d5f3
https://pastila.nl/?001a6151/fb8e29bd7788123a4fa1fbaa13d1906c
https://pastila.nl/?001a6151/21d64c1a454b33614be8b32890deddf3

### CODE REVIEW

**Comment by tavplubix** [comment]
https://github.com/ClickHouse/ClickHouse/pull/45342 should be merged first, this PR should be merged with squash

**Review by rschu1ze — APPROVED** [review]

File: programs/client/Client.cpp [hunk 1]
```diff
@@ -968,7 +968,7 @@ void Client::processOptions(const OptionsDescription & options_description,
             if (external_tables.back().file == "-")
                 ++number_of_external_tables_with_stdin_source;
             if (number_of_external_tables_with_stdin_source > 1)
-                throw Exception("Two or more external tables has stdin (-) set as --file field", ErrorCodes::BAD_ARGUMENTS);
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Two or more external tables has stdin (-) set as --file field");
```
[the remark this answers is not in the document]
Comment (tavplubix) [reply programs/client/Client.cpp:971]: Yep, I will try to forbid it in a separate PR. Need to reformat the remaining 1% of the old messages first...

**Comment by tavplubix** [comment]
ClickHouse special build check - _related_
Integration tests - `test_replicated_users` looks _related_
Integration tests (tsan) [3/6] - https://github.com/ClickHouse/ClickHouse/pull/45072#issuecomment-1399227123
Stress test (debug) - looks like https://github.com/ClickHouse/ClickHouse/issues/45224
Stress test (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/43183

**Comment by tavplubix** [comment]
Another merge conflict...

File: src/Access/KerberosInit.cpp [hunk 2]
```diff
@@ -81,15 +81,15 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
     {
         ret = krb5_cc_resolve(k5.ctx, cache_name.c_str(), &k5.out_cc);
         if (ret)
-            throw Exception("Error in resolving cache" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in resolving cache{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:84]:
```suggestion
            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in resolving cache: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 3]
```diff
@@ -81,15 +81,15 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
...
         LOG_TRACE(log,"Resolved cache");
     }
     else
     {
         // Resolve the default cache and get its type and default principal (if it is initialized).
         ret = krb5_cc_default(k5.ctx, &defcache);
         if (ret)
-            throw Exception("Error while getting default cache" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error while getting default cache{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:92]:
```suggestion
            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error while getting default cache: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 4]
```diff
@@ -99,15 +99,15 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
     // Use the specified principal name.
     ret = krb5_parse_name_flags(k5.ctx, principal.c_str(), 0, &k5.me);
     if (ret)
-        throw Exception("Error when parsing principal name " + principal + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error when parsing principal name {}", principal + fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:102]:
```suggestion
        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error when parsing principal name ({}): {}", principal, fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 5]
```diff
@@ -118,7 +118,7 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
             // Create a new cache to avoid overwriting the initialized default cache.
             ret = krb5_cc_new_unique(k5.ctx, deftype, nullptr, &k5.out_cc);
             if (ret)
-                throw Exception("Error while generating new cache" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+                throw Exception(ErrorCodes::KERBEROS_ERROR, "Error while generating new cache{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:121]:
```suggestion
                throw Exception(ErrorCodes::KERBEROS_ERROR, "Error while generating new cache: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 6]
```diff
@@ -134,24 +134,24 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
 
     ret = krb5_unparse_name(k5.ctx, k5.me, &k5.name);
     if (ret)
-        throw Exception("Error when unparsing name" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error when unparsing name{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:137]:
```suggestion
        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error when unparsing name: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 7]
```diff
@@ -134,24 +134,24 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
...
     LOG_TRACE(log,"Using principal: {}", k5.name);
 
     // Allocate a new initial credential options structure.
     ret = krb5_get_init_creds_opt_alloc(k5.ctx, &options);
     if (ret)
-        throw Exception("Error in options allocation" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in options allocation{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:143]:
```suggestion
        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in options allocation: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 8]
```diff
@@ -134,24 +134,24 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
...
 
     // Resolve keytab
     ret = krb5_kt_resolve(k5.ctx, keytab_file.c_str(), &keytab);
     if (ret)
-        throw Exception("Error in resolving keytab "+keytab_file + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in resolving keytab {}{}", keytab_file, fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:148]:
```suggestion
        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in resolving keytab ({}): {}", keytab_file, fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 9]
```diff
@@ -134,24 +134,24 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
...
     LOG_TRACE(log,"Using keytab: {}", keytab_file);
 
     // Set an output credential cache in initial credential options.
     ret = krb5_get_init_creds_opt_set_out_ccache(k5.ctx, options, k5.out_cc);
     if (ret)
-        throw Exception("Error in setting output credential cache" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in setting output credential cache{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:154]:
```suggestion
        throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in setting output credential cache: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 10]
```diff
@@ -165,7 +165,7 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
         // Request KDC for an initial credentials using keytab.
         ret = krb5_get_init_creds_keytab(k5.ctx, &my_creds, k5.me, keytab, 0, nullptr, options);
         if (ret)
-            throw Exception("Error in getting initial credentials" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in getting initial credentials{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:168]:
```suggestion
            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error in getting initial credentials: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 11]
```diff
@@ -175,7 +175,7 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
         // Initialize a credential cache. Destroy any existing contents of cache and initialize it for the default principal.
         ret = krb5_cc_initialize(k5.ctx, k5.out_cc, k5.me);
         if (ret)
-            throw Exception("Error when initializing cache" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error when initializing cache{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:178]:
```suggestion
            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error when initializing cache: {}", fmtError(ret));
```

File: src/Access/KerberosInit.cpp [hunk 12]
```diff
@@ -189,7 +189,7 @@ void KerberosInit::init(const String & keytab_file, const String & principal, co
         // Make a credential cache the primary cache for its collection.
         ret = krb5_cc_switch(k5.ctx, k5.out_cc);
         if (ret)
-            throw Exception("Error while switching to new cache" + fmtError(ret), ErrorCodes::KERBEROS_ERROR);
+            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error while switching to new cache{}", fmtError(ret));
```
Comment (nikitamikhaylov) [inline src/Access/KerberosInit.cpp:192]:
```suggestion
            throw Exception(ErrorCodes::KERBEROS_ERROR, "Error while switching to new cache: {}", fmtError(ret));
```

**Review by nikitamikhaylov — COMMENTED** [review]
I didn't read all the changes carefully, but looks Okay.

**Comment by tavplubix** [comment]
Fast test - OK
I will merge it optimistically (until another merge conflict appeared) and revert if some tests fail (but all tests should be ok according to previous commits).
Thanks for the suggestions, will add them in a subsequent PR.

**Pull request merged**